### ISSUE TITLE
'overlay' filesystem support being missing is permanently cached (until reboot?)

### ISSUE THREAD

**Issue by srcshelton** [containers/podman#12212]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

(This follows from #1516)

I have a system which tries to start `podman` on boot, which I've just updated to a new kernel.  Something seems to have changed, but whilst the `overlay` module was previous being auto-loaded is no longer is... but this I can fix.

The issue is that if `podman` is *ever* started without overlay support being present, then it will cache this status and continue to report that overlay support is unavailable even when it is.  And if `podman` is started on boot, this means that this is maintained for the entire session or until the magic file is removed.

Admittedly, if the `overlay` module is manually loaded on boot then this would be resolved in the next session (so long as the status file is removed on shutdown?) but this seems an unhelpful and non-obvious way to do things: if overlay support has _become_ available then why not clear the error status?

(I've been using `podman` for some time now and this almost had me downgrading the kernel to figure out what could have changed!  I'm still not entirely sure how `overlay` support was present/autoloaded before and isn't now, although it could have been moved to a module from being built-in in the upstream kernel update...?)

```
$ sudo podman info ; echo ; sudo ls -l /storage/containers/podman/run/overlay/overlay-false ; echo ; sudo cat /storage/containers/podman/run/overlay/overlay-false ; echo ; sudo rm /storage/containers/podman/run/overlay/overlay-false ; sudo podman version
Error: kernel does not support overlay fs: 'overlay' is not supported over extfs at "/storage/containers/podman/storage/overlay": backing file system is unsupported for this graph driver

-rw-r--r-- 1 root sys 179 Nov  7 16:27 /storage/containers/podman/run/overlay/overlay-false

kernel does not support overlay fs: 'overlay' is not supported over extfs at "/storage/containers/podman/storage/overlay": backing file system is unsupported for this graph driver
Version:      3.4.1
API Version:  3.4.1
Go Version:   go1.17.2
Git Commit:   a6493ae6904a5b7dde2395a374cb77fc45c8a3bc
Built:        Sun Oct 24 03:43:26 2021
OS/Arch:      linux/arm64
```

**Output of `podman info --debug`:**

```
host:
  arch: arm64
  buildahVersion: 1.23.1
  cgroupControllers:
  - cpuset
  - cpu
  - io
  - memory
  - pids
  cgroupManager: cgroupfs
  cgroupVersion: v2
  conmon:
    package: app-emulation/conmon-2.0.30
    path: /usr/bin/conmon
    version: 'conmon version 2.0.30, commit: 2792c16f4436f1887a7070d9ad99d9c29742f38a'
  cpus: 4
  distribution:
    distribution: gentoo
    version: unknown
  eventLogger: file
  hostname: turnpike
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 5.10.63-v8+
  linkmode: dynamic
  logDriver: k8s-file
  memFree: 7170617344
  memTotal: 8191483904
  ociRuntime:
    name: crun
    package: app-emulation/crun-1.2
    path: /usr/bin/crun
    version: |-
      crun version 1.2
      commit: 4f6c8e0583c679bfee6a899c05ac6b916022561b
      spec: 1.0.0
      +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
  os: linux
  remoteSocket:
    path: /run/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_AUDIT_WRITE,CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_MKNOD,CAP_NET_BIND_SERVICE,CAP_NET_RAW,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: false
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:
    executable: ""
    package: ""
    version: ""
  swapFree: 8589930496
  swapTotal: 8589930496
  uptime: 11h 39m 37.61s (Approximately 0.46 days)
plugins:
  log:
  - k8s-file
  - none
  network:
  - bridge
  - macvlan
  volume:
  - local
registries:
  localhost:5000:
    Blocked: false
    Insecure: true
    Location: localhost:5000
    MirrorByDigestOnly: false
    Mirrors: []
    Prefix: localhost:5000
  search:
  - docker.io
  - docker.pkg.github.com
  - quay.io
  - public.ecr.aws
store:
  configFile: /etc/containers/storage.conf
  containerStore:
    number: 0
    paused: 0
    running: 0
    stopped: 0
  graphDriverName: overlay
  graphOptions:
    overlay.mountopt: nodev
  graphRoot: /storage/containers/podman/storage
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 0
  runRoot: /storage/containers/podman/run
  volumePath: /storage/containers/podman/volumes
version:
  APIVersion: 3.4.1
  Built: 1635043406
  BuiltTime: Sun Oct 24 03:43:26 2021
  GitCommit: a6493ae6904a5b7dde2395a374cb77fc45c8a3bc
  GoVersion: go1.17.2
  OsArch: linux/arm64
  Version: 3.4.1

```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/master/troubleshooting.md)**

Yes

**Comment by rhatdan** [comment]
@giuseppe @rhvgoyal PTAL
Shouldn't the overlay driver be autoloaded?

**Comment by giuseppe** [comment]
this is a configuration issue, since `runRoot` is supposed to be on a volatile file system like tmpfs that doesn't survive a reboot.

Please change the `runRoot` to be on tmpfs.

We do not have a runtime check since there are use cases when this is fine, e.g., running in a nested container

**Comment by lafleurdeboum** [comment]
I'm half-necro-bumping this issue because it was the first hit when I searched for the string

    podman 'overlay' is not supported over extfs

Just wanted to add that this may also happen if your kernel was not configured with `CONFIG_OVERLAY_FS`.
There is a workaround mentioned on [SO](https://unix.stackexchange.com/a/689182), which is using the `fuse-overlayfs` package. You then need to modify the `mount_program` option, eg in `/etc/containers/storage.conf`.

Hope I'm not polluting the bug tracker. I tried to keep it short anyway.
